### PR TITLE
build: enable PIE by default

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:

### CODE REVIEW

**Comment by lizan** [comment]
@PiotrSikora @agl @davidben is it possible to add `-fPIC` to BoringSSL FIPS build?

**Comment by PiotrSikora** [comment]
As far as I know, we can't since the Security Policy has to be followed to the dot when building module... but @agl is the authority here.

File: .bazelrc [hunk 1]
```diff
@@ -19,6 +19,7 @@ build --action_env=BAZEL_LINKLIBS=-l%:libstdc++.a
 build --action_env=BAZEL_LINKOPTS=-lm
 build --host_javabase=@bazel_tools//tools/jdk:remote_jdk11
 build --javabase=@bazel_tools//tools/jdk:remote_jdk11
+build --copt=-fPIC
```
Comment (htuch) [inline .bazelrc]: Is it possible to add a Python or shell test that verifies the PIE property on either the binary produced by Bazel or when actually running? Given prior discussions, belts-and-braces is ideal if we can do it reasonably.

**Comment by mattklein123** [comment]
Can we add a release note also about this? I think it's useful to let people know that we expect the binary to be runnable under ASLR.

File: test/exe/pie_test.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+ENVOY_BIN=${TEST_SRCDIR}/envoy/source/exe/envoy-static
```
Comment (htuch) [inline test/exe/pie_test.sh]: Nit: use proper shell `""` around path expanding vars.

File: test/exe/pie_test.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+if [[ `uname` == "Darwin" ]]; then
+  echo "Skipping on macOS."
+  exit 0
+fi
+
+if readelf -hW ${ENVOY_BIN} | grep "Type" | grep -o "DYN (Shared object file)"; then
```
Comment (htuch) [inline test/exe/pie_test.sh]: Same here.

File: test/exe/pie_test.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
```
Comment (htuch) [inline test/exe/pie_test.sh]: Nit: `set -e`

File: test/exe/pie_test.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  exit 0
+fi
+
+if readelf -hW ${ENVOY_BIN} | grep "Type" | grep -o "DYN (Shared object file)"; then
+  echo "${ENVOY_BIN} is a PIE!"
+  exit 0
+fi
+
+echo "${ENVOY_BIN} is not a PIE!"
+exit 1
```
Comment (htuch) [inline test/exe/pie_test.sh]: I assume you verified this with before/after binaries..

Comment (lizan) [reply test/exe/pie_test.sh]: yes

**Comment by nareddyt** [comment]
HI @lizan, could you explain the `-fPIC` copt you added to `.bazelrc`? I get what the copt does, but I don't understand why it's needed to build now.

**Comment by lizan** [comment]
@nareddyt to link a binary with `-pie` all objects (`.o` files) has to be compiled with `-fPIC`. --copt is propagating all the way down to dependencies including foreign_cc cmake rules.

**Pull request merged**